aptos-move/block-executor/src/view.rs

```diff
@@ -627,11 +768,149 @@ impl<'a, T: Transaction, X: Executable> SequentialState<'a, T, X> {
+                if let UnknownOrLayout::Known(layout) = layout {
+                    if let ValueWithLayout::RawFromStorage(v) = value {
+                        match patch_base_value(v.as_ref(), layout) {
+                            Ok(patched_value) => {
+                                let exchanged_value = ValueWithLayout::Exchanged(
+                                    Arc::new(patched_value.clone()),
+                                    layout.cloned().map(Arc::new),
+                                );
+                                self.unsync_map
+                                    .set_base_value(key.clone(), exchanged_value.clone());
+
+                                // sequential execution doesn't need to worry about concurrent change going through.
+                                value = exchanged_value;
+                            },
+                            Err(_) => {
+                                *self.incorrect_use.borrow_mut() = true;
+                                error!("Unsync map couldn't patch base value");
+                                return ReadResult::HaltSpeculativeExecution(
+                                    "Unsync map couldn't patch base value".to_string(),
+                                );
+                            },
+                        }
+                    }
+                }
+
+                if let Some(ret) = ReadResult::from_value_with_layout(value, target_kind.clone()) {
+                    if target_kind == ReadKind::Value {
+                        if let ReadResult::Value(v, l) = &ret {
+                            if v.is_some() && l.is_some() {
+                                self.resource_with_layout_read_set
+                                    .borrow_mut()
+                                    .insert(key.clone());
+                            }
+                        }
+                    }
+                    ret
+                } else {
+                    *self.incorrect_use.borrow_mut() = true;
+                    error!(
+                        "Unsync map has RawFromStorage value type, while we are requesting value"
+                    );
+                    ReadResult::HaltSpeculativeExecution(
+                        "Unsync map has RawFromStorage value type, while we are requesting value"
+                            .to_string(),
+                    )
+                }
+            },
+            None => ReadResult::Uninitialized,
+        }
+    }
+}
+
+impl<'a, T: Transaction, X: Executable> ResourceGroupState<T> for SequentialState<'a, T, X> {
+    fn set_raw_group_base_values(&self, group_key: T::Key, base_values: Vec<(T::Tag, T::Value)>) {
+        self.unsync_map
+            .set_group_base_values(group_key.clone(), base_values);
+    }
+
+    fn read_cached_group_tagged_data(
+        &self,
+        _txn_idx: TxnIndex,
+        group_key: &T::Key,
+        resource_tag: &T::Tag,
+        maybe_layout: Option<&MoveTypeLayout>,
```

Comment:
should we also make this UnknownOrLayout? might allow to re-use some code, and in the future we may also add exists for resources in groups too

Reply:
this is intentional.

UnknownOrLayout has one more value:
- Unknown
- Known(None)
- Known(Some(layout))

we cannot have Unknown here, so using UnknownOrLayout is not right.

in non-group case, we have a kind parameter, and function receives layout always when kind is value, and can have a missing layout otherwise.

Here we don't have kind, and the only two "kinds" - size and value are separate functions (as size refers to the whole group, vs value which refers to a particular tag)

longterm, maybe we should fully align those two reads, and cleanup more, but at this point this is the correct thing to do

Reply:
yeah, I guess longer term, I can see us having exists for resources in the group too, and that one wouldn't have a kind.

Reply:
We can type alias Option<&MoveTypeLayout> to KnownLayout to make things more uniform I guess